klogd.8
.\" Copyright 1994-6 Dr. Greg Wettstein, Enjellic Systems Development.
.\" Copyright 1997-2007 Martin Schulze <joey@infodrom.org>
.\" Thu Feb 17 2000: Chris Wing: Unprivileged klogd feature
.\" May be distributed under the GNU General Public License
.\"
.TH KLOGD 8 "27 May 2007" "Version 1.5" "Linux System Administration"
.SH NAME
klogd \- Kernel Log Daemon
.SH SYNOPSIS
.B klogd
.RB [ " \-c "
.I n
]
.RB [ " \-d " ]
.RB [ " \-f "
.I fname
]
.RB [ " \-u "
.I username
]
.RB [ " \-j "
.I chroot_dir
]
.RB [ " \-iI " ]
.RB [ " \-n " ]
.RB [ " \-o " ]
.RB [ " \-p " ]
.RB [ " \-s " ]
.RB [ " \-k "
.I fname
]
.RB [ " \-v " ]
.RB [ " \-x " ]
.RB [ " \-2 " ]
.SH DESCRIPTION
.B klogd
is a system daemon which intercepts and logs Linux kernel
messages.
.SH OPTIONS
.TP
.BI "\-c " n
Sets the default log level of console messages to \fIn\fR.
.TP
.B "\-d"
Enable debugging mode.  This will generate \fBLOTS\fR of output to
stderr.
.TP
.BI "\-f " file
Log messages to the specified filename rather than to the syslog facility.
.TP
.BI "\-u " username
Tells klogd to become the specified user and drop root privileges before
starting logging.
.TP
.BI "\-j " chroot_dir
Tells klogd to
.BR chroot (2)
into this directory after initializing.
This option is only valid if the \-u option is also used to run klogd
without root privileges.
Note that the use of this option will prevent \-i and \-I from working
unless you set up the chroot directory in such a way that klogd can still
read the kernel module symbols.
.TP
.BI "\-i \-I"
Signal the currently executing klogd daemon.  Both of these switches control
the loading/reloading of symbol information.  The \-i switch signals the
daemon to reload the kernel module symbols.  The \-I switch signals for a
reload of both the static kernel symbols and the kernel module symbols.
.TP
.B "\-n"
Avoid auto-backgrounding.  This is needed especially if the
.B klogd
is started and controlled by 
.BR init (8).
.TP
.B "\-o"
Execute in 'one\-shot' mode.  This causes \fBklogd\fP to read and log
all the messages that are found in the kernel message buffers.  After
a single read and log cycle the daemon exits.
.TP
.B "\-p"
Enable paranoia.  This option controls when klogd loads kernel module symbol
information.  Setting this switch causes klogd to load the kernel module
symbol information whenever an Oops string is detected in the kernel message
stream.
.TP
.B "\-s"
Force \fBklogd\fP to use the system call interface to the kernel message
buffers.
.TP
.BI "\-k " file
Use the specified file as the source of kernel symbol information.
.TP
.B "\-v"
Print version and exit.
.TP
.B "\-x"
Omits EIP translation and therefore doesn't read the System.map file.
.TP
.B "\-2"
When symbols are expanded, print the line twice.  Once with addresses
converted to symbols, once with the raw text.  This allows external
programs such as ksymoops do their own processing on the original
data.
.SH OVERVIEW
The functionality of klogd has been typically incorporated into other
versions of syslogd but this seems to be a poor place for it.  In the
modern Linux kernel a number of kernel messaging issues such as
sourcing, prioritization and resolution of kernel addresses must be
addressed.  Incorporating kernel logging into a separate process
offers a cleaner separation of services.

In Linux there are two potential sources of kernel log information: the 
.I /proc
file system and the syscall (sys_syslog) interface, although
ultimately they are one and the same.  Klogd is designed to choose
whichever source of information is the most appropriate.  It does this
by first checking for the presence of a mounted 
.I /proc
file system.  If this is found the 
.I /proc/kmsg
file is used as the source of kernel log
information.  If the proc file system is not mounted 
.B klogd
uses a
system call to obtain kernel messages.  The command line switch
.RB ( "\-s" )
can be used to force klogd to use the system call interface as its
messaging source.

If kernel messages are directed through the 
.BR syslogd " daemon the " klogd
daemon, as of version 1.1, has the ability to properly prioritize
kernel messages.  Prioritization of the kernel messages was added to it
at approximately version 0.99pl13 of the kernel.  The raw kernel messages
are of the form:
.IP
\<[0\-7]\>Something said by the kernel.
.PP
The priority of the kernel message is encoded as a single numeric
digit enclosed inside the <> pair.  The definitions of these values is
given in the kernel include file kernel.h.  When a message is received
from the kernel the klogd daemon reads this priority level and assigns
the appropriate priority level to the syslog message.  If file output
(\fB-f\fR) is used the prioritization sequence is left pre\-pended to the
kernel message.

The klogd daemon can also be used in a 'one\-shot' mode for reading the
kernel message buffers.  One shot mode is selected by specifying the
\fB\-o\fR switch on the command line.  Output will be directed to either the
syslogd daemon or to an alternate file specified by the \fB-f\fR switch.
.IP
For example, to read all the kernel messages after a system
boot and record them in a file called krnl.msg the following
command would be given.
.IP
.nf
	klogd -o -f ./krnl.msg
.fi
.SH KERNEL ADDRESS RESOLUTION
If the kernel detects an internal error condition a general protection
fault will be triggered.  As part of the GPF handling procedure the
kernel prints out a status report indicating the state of the
processor at the time of the fault.  Included in this display are the
contents of the microprocessor's registers, the contents of the kernel
stack and a tracing of what functions were being executed at the time
of the fault.

This information is
.B EXTREMELY IMPORTANT
in determining what caused the internal error condition.  The
difficulty comes when a kernel developer attempts to analyze this
information.  The raw numeric information present in the protection
fault printout is of very little use to the developers.  This is due
to the fact that kernels are not identical and the addresses of
variable locations or functions will not be the same in all kernels.
In order to correctly diagnose the cause of failure a kernel developer
needs to know what specific kernel functions or variable locations
were involved in the error.

As part of the kernel compilation process a listing is created which
specified the address locations of important variables and function in
the kernel being compiled.  This listing is saved in a file called
System.map in the top of the kernel directory source tree.  Using this
listing a kernel developer can determine exactly what the kernel was
doing when the error condition occurred.

The process of resolving the numeric addresses from the protection
fault printout can be done manually or by using the
.B ksymoops
program which is included in the kernel sources.

As a convenience
.B klogd
will attempt to resolve kernel numeric addresses to their symbolic
forms if a kernel symbol table is available at execution time.  If you
require the original address of the symbol, use the
.B -2
switch to preserve the numeric address.  A
symbol table may be specified by using the \fB\-k\fR switch on the
command line.  If a symbol file is not explicitly specified the
following filenames will be tried:

.nf
.I /boot/System.map
.I /System.map
.I /usr/src/linux/System.map
.fi

Version information is supplied in the system maps as of kernel
1.3.43.  This version information is used to direct an intelligent
search of the list of symbol tables.  This feature is useful since it
provides support for both production and experimental kernels.

For example a production kernel may have its map file stored in
/boot/System.map.  If an experimental or test kernel is compiled with
the sources in the 'standard' location of /usr/src/linux the system
map will be found in /usr/src/linux/System.map.  When klogd starts
under the experimental kernel the map in /boot/System.map will be
bypassed in favor of the map in /usr/src/linux/System.map.

Modern kernels as of 1.3.43 properly format important kernel addresses
so that they will be recognized and translated by klogd.  Earlier
kernels require a source code patch be applied to the kernel sources.
This patch is supplied with the sysklogd sources.

The process of analyzing kernel protections faults works very well
with a static kernel.  Additional difficulties are encountered when
attempting to diagnose errors which occur in loadable kernel modules.
Loadable kernel modules are used to implement kernel functionality in
a form which can be loaded or unloaded at will.  The use of loadable
modules is useful from a debugging standpoint and can also be useful
in decreasing the amount of memory required by a kernel.

The difficulty with diagnosing errors in loadable modules is due to
the dynamic nature of the kernel modules.  When a module is loaded the
kernel will allocate memory to hold the module, when the module is
unloaded this memory will be returned back to the kernel.  This
dynamic memory allocation makes it impossible to produce a map file
which details the addresses of the variable and functions in a kernel
loadable module.  Without this location map it is not possible for a
kernel developer to determine what went wrong if a protection fault
involves a kernel module.

.B klogd
has support for dealing with the problem of diagnosing protection
faults in kernel loadable modules.  At program start time or in
response to a signal the daemon will interrogate the kernel for a
listing of all modules loaded and the addresses in memory they are
loaded at.  Individual modules can also register the locations of
important functions when the module is loaded.  The addresses of these
exported symbols are also determined during this interrogation
process.

When a protection fault occurs an attempt will be made to resolve
kernel addresses from the static symbol table.  If this fails the
symbols from the currently loaded modules are examined in an attempt
to resolve the addresses.  At the very minimum this allows klogd to
indicate which loadable module was responsible for generating the
protection fault.  Additional information may be available if the
module developer chose to export symbol information from the module.

Proper and accurate resolution of addresses in kernel modules requires
that
.B klogd
be informed whenever the kernel module status changes.  The
.B \-i
and
.B \-I
switches can be used to signal the currently executing daemon that
symbol information be reloaded.  Of most importance to proper
resolution of module symbols is the
.B \-i
switch.  Each time a kernel module is loaded or removed from the
kernel the following command should be executed:

.nf
.I klogd \-i
.fi

The
.B \-p
switch can also be used to insure that module symbol information is up
to date.  This switch instructs
.B klogd
to reload the module symbol information whenever a protection fault
is detected.  Caution should be used before invoking the program in
\&'paranoid\&' mode.  The stability of the kernel and the operating
environment is always under question when a protection fault occurs.
Since the klogd daemon must execute system calls in order to read the
module symbol information there is the possibility that the system may
be too unstable to capture useful information.  A much better policy
is to insure that klogd is updated whenever a module is loaded or
unloaded.  Having uptodate symbol information loaded increases the
probability of properly resolving a protection fault if it should occur.

Included in the sysklogd source distribution is a patch to the
modules-2.0.0 package which allows the
.B insmod,
.B rmmod
and
.B modprobe
utilities to automatically signal
.B klogd
whenever a module is inserted or removed from the kernel.  Using this
patch will insure that the symbol information maintained in klogd is
always consistent with the current kernel state.
.SH CONSOLE LOG LEVEL
The
.B klogd
daemon allows the ability to alter the presentation of
kernel messages to the system console.  Consequent with the
prioritization of kernel messages was the inclusion of default
messaging levels for the kernel.  In a stock kernel the the default
console log level is set to 7.  Any messages with a priority level
numerically lower than 7 (higher priority) appear on the console.

Messages of priority level 7 are considered to be 'debug' messages and
will thus not appear on the console.  Many administrators,
particularly in a multi\-user environment, prefer that all kernel
messages be handled by klogd and either directed to a file or to
the syslogd daemon.  This prevents 'nuisance' messages such as line
printer out of paper or disk change detected from cluttering the
console.

When
.B \-c
is given on the commandline the
.B klogd
daemon will execute a system call to inhibit all kernel messages from
being displayed on the console.  Former versions always issued this
system call and defaulted to all kernel messages except for panics.
This is handled differently currently so
.B klogd
doesn't need to set this value anymore.  The
argument given to the \fB\-c\fR switch specifies the priority level of
messages which will be directed to the console.  Note that messages of
a priority value LOWER than the indicated number will be directed to
the console.
.IP
For example, to have the kernel display all messages with a
priority level of 3
.BR "" ( KERN_ERR )
or more severe the following
command would be executed:
.IP
.nf
	klogd \-c 4
.fi
.PP
The definitions of the numeric values for kernel messages are given in
the file 
.IR kernel.h " which can be found in the " /usr/include/linux
directory if the kernel sources are installed.  These values parallel
the syslog priority values which are defined in the file 
.IR syslog.h " found in the " /usr/include/sys " sub\-directory."

The console log level is usually configured with the
.BR sysctl (8)
program, directly or via its configuration file
.IR /etc/sysctl.conf .
In this file the following line 
.IP
.nf
kernel.printk = 4 4 1 7
.fi
.PP
corresponds to the sampe setting above.
.SH SIGNAL HANDLING
The 
.B klogd
will respond to eight signals:
.BR SIGHUP ", " SIGINT ", " SIGKILL ", " SIGTERM ", " SIGTSTP ", "
.BR SIGUSR1 ", "SIGUSR2 " and " SIGCONT ".  The"
.BR SIGINT ", " SIGKILL ", " SIGTERM " and " SIGHUP
signals will cause the daemon to close its kernel log sources and
terminate gracefully.

The 
.BR SIGTSTP " and " SIGCONT
signals are used to start and stop kernel logging.  Upon receipt of a 
.B SIGTSTP
signal the daemon will close its
log sources and spin in an idle loop.  Subsequent receipt of a 
.B SIGCONT
signal will cause the daemon to go through its initialization sequence
and re-choose an input source.  Using
.BR SIGSTOP " and " SIGCONT
in combination the kernel log input can be re-chosen without stopping and
restarting the daemon.  For example if the \fI/proc\fR file system is to be
un-mounted the following command sequence should be used:
.PP
.PD 0
.TP
	# kill -TSTP pid
.TP
	# umount /proc
.TP
	# kill -CONT pid
.PD
.PP
Notations will be made in the system logs with 
.B LOG_INFO
priority
documenting the start/stop of logging.

The 
.BR SIGUSR1 " and " SIGUSR2
signals are used to initiate loading/reloading of kernel symbol information.
Receipt of the
.B SIGUSR1
signal will cause the kernel module symbols to be reloaded.  Signaling the
daemon with
.B SIGUSR2
will cause both the static kernel symbols and the kernel module symbols to
be reloaded.

Provided that the System.map file is placed in an appropriate location the
signal of generally greatest usefulness is the
.B SIGUSR1
signal.  This signal is designed to be used to signal the daemon when kernel
modules are loaded/unloaded.  Sending this signal to the daemon after a
kernel module state change will insure that proper resolution of symbols will
occur if a protection fault occurs in the address space occupied by a kernel
module.
.SH FILES
.PD 0
.TP
.I /proc/kmsg
One Source for kernel messages
.B klogd
.TP
.I /var/run/klogd.pid
The file containing the process id of 
.B klogd
.TP
.I /boot/System.map, /System.map, /usr/src/linux/System.map
Default locations for kernel system maps.
.PD
.SH BUGS
Probably numerous.  Well formed context diffs appreciated.
.SH AUTHORS
The kernel log daemon
.B klogd
was originally written by Steve Lord <lord@cray.com>, Greg Wettstein
made major improvements.  Martin Schulze <joey@infodrom.org> fixed
some bugs and took over maintenance.


klogd.c
/*
    klogd.c - main program for Linux kernel log daemon.
    Copyright (c) 1995  Dr. G.W. Wettstein <greg@wind.rmcc.com>

    This file is part of the sysklogd package, a kernel and system log daemon.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/fcntl.h>
#include <sys/stat.h>
#if !defined(__GLIBC__)
#include <linux/time.h>
#endif /* __GLIBC__ */
#include <stdarg.h>
#include <paths.h>
#include <stdlib.h>
#include <pwd.h>
#include <grp.h>
#include "klogd.h"
#include "pidfile.h"
#include "version.h"

#define __LIBRARY__
#include <linux/unistd.h>
#if !defined(__GLIBC__)
# define __NR_ksyslog __NR_syslog
_syscall3(int,ksyslog,int, type, char *, buf, int, len);
#else
#include <sys/klog.h>
#define ksyslog klogctl
#endif

#ifndef _PATH_DEVNULL
#define _PATH_DEVNULL	"/dev/null"
#endif

#define LOG_BUFFER_SIZE 4096
#define LOG_LINE_LENGTH 1000

#if defined(FSSTND)
static char	*PidFile = _PATH_VARRUN "klogd.pid";
#else
static char	*PidFile = "/etc/klogd.pid";
#endif

static int	kmsg,
		change_state = 0,
		caught_TSTP = 0,
		console_log_level = -1;

static int	use_syscall = 0,
		one_shot = 0,
		no_fork = 0;	/* don't fork - don't run in daemon mode */

static char	log_buffer[LOG_BUFFER_SIZE];

static FILE *output_file = (FILE *) 0;

static enum LOGSRC {none, proc, kernel} logsrc;

int debugging = 0;

static char *server_user = NULL;
static char *chroot_dir = NULL;
static int log_flags = 0;

/* Function prototypes. */
extern int ksyslog(int type, char *buf, int len);
static void CloseLogSrc(void);
static void Terminate(void);
static void SignalDaemon(int);
static void ChangeLogging(void);
static enum LOGSRC GetKernelLogSrc(void);
static void LogLine(char *ptr, int len);
static void LogKernelLine(void);
static void LogProcLine(void);
extern int main(int argc, char *argv[]);


static void CloseLogSrc(void)
{
	/* Shutdown the log sources. */
	switch ( logsrc )
	{
		case kernel:
			ksyslog(0, 0, 0);
			Syslog(LOG_INFO, "Kernel logging (ksyslog) stopped.");
			break;
		case proc:
			close(kmsg);
			Syslog(LOG_INFO, "Kernel logging (proc) stopped.");
			break;
		case none:
			break;
	}

	if ( output_file != (FILE *) 0 )
		fflush(output_file);
	return;
}


/*
 * Signal handler to terminate the parent process.
 */
static void doexit(int sig)
{
	exit (0);
}

static void restart(int sig)
{
	signal(SIGCONT, restart);
	change_state = 1;
	caught_TSTP = 0;
	return;
}


static void stop_logging(int sig)
{
	signal(SIGTSTP, stop_logging);
	change_state = 1;
	caught_TSTP = 1;
	return;
}


static void stop_daemon(int sig)
{
	Terminate();
}


static void reload_daemon(int sig)
{
	change_state = 1;

	if ( sig == SIGUSR2 )
		signal(SIGUSR2, reload_daemon);
	else
		signal(SIGUSR1, reload_daemon);
}


static void Terminate(void)
{
	CloseLogSrc();
	Syslog(LOG_INFO, "Kernel log daemon terminating.");
	sleep(1);
	if ( output_file != (FILE *) 0 )
		fclose(output_file);
	closelog();
	(void) remove_pid(PidFile);
	exit(1);
}

static void SignalDaemon(int sig)
{
	int pid = check_pid(PidFile);
	kill(pid, sig);
}


static void ChangeLogging(void)
{
	/* Indicate that something is happening. */
	Syslog(LOG_INFO, "klogd %s.%s, ---------- state change ----------\n", \
	       VERSION, PATCHLEVEL);

	/* Stop kernel logging. */
	if ( caught_TSTP == 1 )
	{
		CloseLogSrc();
		logsrc = none;
		change_state = 0;
		return;
	}

	/*
	 * The rest of this function is responsible for restarting
	 * kernel logging after it was stopped.
	 *
	 * In the following section we make a decision based on the
	 * kernel log state as to what is causing us to restart.  Somewhat
	 * groady but it keeps us from creating another static variable.
	 */
	if ( logsrc != none )
	{
		Syslog(LOG_INFO, "Kernel logging re-started after SIGSTOP.");
		change_state = 0;
		return;
	}

	/* Restart logging. */
	logsrc = GetKernelLogSrc();
	change_state = 0;
	return;
}


static enum LOGSRC GetKernelLogSrc(void)
{
	struct stat sb;

	/* Set level of kernel console messaging.. */
	if ( (console_log_level != -1)
	&& (ksyslog(8, NULL, console_log_level) < 0) && \
	     (errno == EINVAL) )
	{
		/*
		 * An invalid arguement error probably indicates that
		 * a pre-0.14 kernel is being run.  At this point we
		 * issue an error message and simply shut-off console
		 * logging completely.
		 */
		Syslog(LOG_WARNING, "Cannot set console log level - disabling "
		       "console output.");
	}

	/*
	 * First do a stat to determine whether or not the proc based
	 * file system is available to get kernel messages from.
	 */
	if (!server_user &&
	    (use_syscall ||
	    ((stat(_PATH_KLOG, &sb) < 0) && (errno == ENOENT))))
	{
		/* Initialize kernel logging. */
		ksyslog(1, NULL, 0);
		Syslog(LOG_INFO, "klogd %s.%s, log source = ksyslog "
		       "started.", VERSION, PATCHLEVEL);
		return(kernel);
	}

	if ( (kmsg = open(_PATH_KLOG, O_RDONLY)) < 0 )
	{
		fprintf(stderr, "klogd: Cannot open proc file system, " \
			"%d - %s.\n", errno, strerror(errno));
		ksyslog(7, NULL, 0);
		exit(1);
	}
	Syslog(LOG_INFO, "klogd %s.%s, log source = %s started.", \
	       VERSION, PATCHLEVEL, _PATH_KLOG);
	return(proc);
}

void Syslog(int priority, char *fmt, ...)
{
	va_list ap;
	char *argl;

	if ( debugging )
	{
		fputs("Logging line:\n", stderr);
		fprintf(stderr, "\tLine: %s\n", fmt);
		fprintf(stderr, "\tPriority: %d\n", priority);
	}

	/* Handle output to a file. */
	if ( output_file != (FILE *) 0 )
	{
		va_start(ap, fmt);
		vfprintf(output_file, fmt, ap);
		va_end(ap);
		fputc('\n', output_file);
		fflush(output_file);
		if (!one_shot)
			fsync(fileno(output_file));
		return;
	}

	/* Output using syslog. */
	if (!strcmp(fmt, "%s"))
	{
		va_start(ap, fmt);
		argl = va_arg(ap, char *);
		if (argl[0] == '<' && argl[1] && argl[2] == '>')
		{
			switch ( argl[1] )
			{
			case '0':
				priority = LOG_EMERG;
				break;
			case '1':
				priority = LOG_ALERT;
				break;
			case '2':
				priority = LOG_CRIT;
				break;
			case '3':
				priority = LOG_ERR;
				break;
			case '4':
				priority = LOG_WARNING;
				break;
			case '5':
				priority = LOG_NOTICE;
				break;
			case '6':
				priority = LOG_INFO;
				break;
			case '7':
			default:
				priority = LOG_DEBUG;
			}
			argl += 3;
		}
		syslog(priority, fmt, argl);
		va_end(ap);
		return;
	}

	va_start(ap, fmt);
	vsyslog(priority, fmt, ap);
	va_end(ap);

	return;
}


/*
 *     Copy characters from ptr to line until a char in the delim
 *     string is encountered or until min( space, len ) chars have
 *     been copied.
 *
 *     Returns the actual number of chars copied.
 */
static int copyin( char *line,      int space,
                   const char *ptr, int len,
                   const char *delim )
{
	auto int i;
	auto int count;

	count = len < space ? len : space;

	for(i = 0; i < count && !strchr(delim, *ptr); i++ ) {
		*line++ = *ptr++;
	}

	return i;
}

/*
 * Messages are separated by "\n".  Messages longer than
 * LOG_LINE_LENGTH are broken up.
 */
static void LogLine(char *ptr, int len)
{
	static char line_buff[LOG_LINE_LENGTH];
	static char *line = line_buff;
	static int space = sizeof(line_buff)-1;

	int delta = 0; /* number of chars copied        */

	while( len > 0 )
	{
		if( space == 0 )    /* line buffer is full */
		{
			/*
			 ** Line too long.  Start a new line.
			 */
			*line = 0;   /* force null terminator */

			if ( debugging )
			{
				fputs("Line buffer full:\n", stderr);
				fprintf(stderr, "\tLine: %s\n", line);
			}

			Syslog( LOG_INFO, "%s", line_buff );
			line  = line_buff;
			space = sizeof(line_buff)-1;
		}

		delta = copyin(line, space, ptr, len, "\n");
		line  += delta;
		ptr   += delta;
		space -= delta;
		len   -= delta;

		if( space == 0 || len == 0 )
			; /* full line_buff or end of input buffer */
		else if( *ptr == '\0' )  /* zero byte */
		{
			ptr++;	/* skip zero byte */
			space -= 1;
			len   -= 1;
		}
		else if( *ptr == '\n' )  /* newline */
		{
			ptr++;	/* skip newline */
			space -= 1;
			len   -= 1;

			*line = 0;  /* force null terminator */

			Syslog( LOG_INFO, "%s", line_buff );
			line  = line_buff;
			space = sizeof(line_buff)-1;
		}
	}
}


static void LogKernelLine(void)
{
	auto int rdcnt;

	/*
	 * Zero-fill the log buffer.  This should cure a multitude of
	 * problems with klogd logging the tail end of the message buffer
	 * which will contain old messages.  Then read the kernel log
	 * messages into this fresh buffer.
	 */
	memset(log_buffer, '\0', sizeof(log_buffer));
	if ( (rdcnt = ksyslog(2, log_buffer, sizeof(log_buffer)-1)) < 0 )
	{
		if ( errno == EINTR )
			return;
		fprintf(stderr, "klogd: Error return from sys_sycall: " \
			"%d - %s\n", errno, strerror(errno));
	}
	else
		LogLine(log_buffer, rdcnt);
	return;
}


static void LogProcLine(void)
{
	auto int rdcnt;

	/*
	 * Zero-fill the log buffer.  This should cure a multitude of
	 * problems with klogd logging the tail end of the message buffer
	 * which will contain old messages.  Then read the kernel messages
	 * from the message pseudo-file into this fresh buffer.
	 */
	memset(log_buffer, '\0', sizeof(log_buffer));
	if ( (rdcnt = read(kmsg, log_buffer, sizeof(log_buffer)-1)) < 0 )
	{
		int saved_errno = errno;

		if ( errno == EINTR )
			return;
		Syslog(LOG_ERR, "Cannot read proc file system: %d - %s.", \
		       errno, strerror(errno));
		if ( saved_errno == EPERM )
			Terminate();
	}
	else
		LogLine(log_buffer, rdcnt);

	return;
}


static int drop_root(void)
{
	struct passwd *pw;

	if (!(pw = getpwnam(server_user))) return -1;

	if (!pw->pw_uid) return -1;

	if (chroot_dir) {
		if (chdir(chroot_dir)) return -1;
		if (chroot(".")) return -1;
	}

	if (setgroups(0, NULL)) return -1;
	if (setgid(pw->pw_gid)) return -1;
	if (setuid(pw->pw_uid)) return -1;

	return 0;
}


int main(int argc, char *argv[])
{
	auto int	ch,
			use_output = 0;

	auto char	*log_level = (char *) 0,
			*output = (char *) 0;

	pid_t ppid = getpid();
	if (chdir ("/") < 0) {
		fprintf(stderr, "klogd: chdir to / failed: %m");
		exit(1);
	}

	/* Parse the command-line. */
	while ((ch = getopt(argc, argv, "c:df:u:j:iIk:nopsvx2")) != EOF)
		switch((char)ch)
		{
		    case '2':		/* Print lines with symbols twice. */
			break;
		    case 'c':		/* Set console message level. */
			log_level = optarg;
			break;
		    case 'd':		/* Activity debug mode. */
			debugging = 1;
			break;
		    case 'f':		/* Define an output file. */
			output = optarg;
			use_output++;
			break;
		    case 'i':		/* Reload module symbols. */
			SignalDaemon(SIGUSR1);
			return(0);
		    case 'I':
			SignalDaemon(SIGUSR2);
			return(0);
		    case 'j':		/* chroot 'j'ail */
			chroot_dir = optarg;
			log_flags |= LOG_NDELAY;
			break;
		    case 'k':		/* Kernel symbol file. */
			break;
		    case 'n':		/* don't fork */
			no_fork++;
			break;
		    case 'o':		/* One-shot mode. */
			one_shot = 1;
			break;
		    case 'p':
			break;
		    case 's':		/* Use syscall interface. */
			use_syscall = 1;
			break;
		    case 'u':		/* Run as this user */
			server_user = optarg;
			break;
		    case 'v':
			printf("klogd %s.%s\n", VERSION, PATCHLEVEL);
			exit (1);
		    case 'x':
			break;
		}

	if (chroot_dir && !server_user) {
		fputs("'-j' is only valid with '-u'\n", stderr);
		exit(1);
	}

	/* Set console logging level. */
	if ( log_level != (char *) 0 )
	{
		if ( (strlen(log_level) > 1) || \
		     (strchr("12345678", *log_level) == (char *) 0) )
		{
			fprintf(stderr, "klogd: Invalid console logging "
				"level <%s> specified.\n", log_level);
			return(1);
		}
		console_log_level = *log_level - '0';
	}

	/*
	 * The following code allows klogd to auto-background itself.
	 * What happens is that the program forks and the parent quits.
	 * The child closes all its open file descriptors, and issues a
	 * call to setsid to establish itself as an independent session
	 * immune from control signals.
	 *
	 * fork() is only called if it should run in daemon mode, fork is
	 * not disabled with the command line argument and there's no
	 * such process running.
	 */
	if ( (!one_shot) && (!no_fork) )
	{
		if (!check_pid(PidFile))
		{
			signal (SIGTERM, doexit);
			pid_t pid;
			int fl;

			if ( (fl = open(_PATH_DEVNULL, O_RDWR)) < 0 )
			{
				fprintf(stderr, "klogd: %s: %s\n",
				         _PATH_DEVNULL, strerror(errno));
				exit(1);
			}

			if ( (pid = fork()) == -1 )
			{
				fputs("klogd: fork failed.\n", stderr);
				exit(1);
			} else if ( pid == 0 )
			{
				int num_fds = getdtablesize();

				signal (SIGTERM, SIG_DFL);

				/* This is the child closing its file descriptors. */
				if ( dup2(fl, 0) != 0 ||
				     ((!use_output || strcmp(output, "-")) &&
				      dup2(fl, 1) != 1) ||
				     dup2(fl, 2) != 2)
				{
					fputs("klogd: dup2 failed.\n", stderr);
					exit(1);
				}
				for (fl= 3; fl <= num_fds; ++fl)
					close(fl);

				setsid();
			}
			else
			{
				/*
				 * Parent process
				 */
				sleep(300);
				/*
				 * Not reached unless something major went wrong.
				 */
				exit(1);
			}
		}
		else
		{
			fputs("klogd: Already running.\n", stderr);
			exit(1);
		}
	}


	/* tuck my process id away */
	if (!check_pid(PidFile))
	{
		if (!write_pid(PidFile))
			Terminate();
	}
	else
	{
		fputs("klogd: Already running.\n", stderr);
		Terminate();
	}

	/* Signal setups. */
	for (ch= 1; ch < NSIG; ++ch)
		signal(ch, SIG_IGN);
	signal(SIGINT, stop_daemon);
	signal(SIGKILL, stop_daemon);
	signal(SIGTERM, stop_daemon);
	signal(SIGHUP, stop_daemon);
	signal(SIGTSTP, stop_logging);
	signal(SIGCONT, restart);
	signal(SIGUSR1, reload_daemon);
	signal(SIGUSR2, reload_daemon);


	/* Open outputs. */
	if ( use_output )
	{
		if ( strcmp(output, "-") == 0 )
			output_file = stdout;
		else if ( (output_file = fopen(output, "w")) == (FILE *) 0 )
		{
			fprintf(stderr, "klogd: Cannot open output file " \
				"%s - %s\n", output, strerror(errno));
			return(1);
		}
	}
	else
		openlog("kernel", log_flags, LOG_KERN);


	/* Handle one-shot logging. */
	if ( one_shot )
	{
		if ( (logsrc = GetKernelLogSrc()) == kernel )
			LogKernelLine();
		else
			LogProcLine();
		Terminate();
	}

	/* Determine where kernel logging information is to come from. */
#if defined(KLOGD_DELAY)
	sleep(KLOGD_DELAY);
#endif
	logsrc = GetKernelLogSrc();

	if (getpid() != ppid)
		kill (ppid, SIGTERM);

	if (server_user && drop_root()) {
		syslog(LOG_ALERT, "klogd: failed to drop root");
		Terminate();
	}

        /* The main loop. */
	while (1)
	{
		if ( change_state )
			ChangeLogging();
		switch ( logsrc )
		{
			case kernel:
				LogKernelLine();
				break;
			case proc:
				LogProcLine();
				break;
			case none:
				pause();
				break;
		}
	}
}
/*
 * Local variables:
 *  c-indent-level: 8
 *  c-basic-offset: 8
 *  tab-width: 8
 * End:
 */
